src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java

```diff
@@ -0,0 +1,131 @@
+    private static final Logger logger = LoggerFactory.getLogger(IndexSearchController.class);
+    @FXML
+    public Pane resultContainer;
+    @FXML
+    public HBox loader;
+    @FXML
+    private Pane indexSearchRoot;
+    @FXML
+    JFXTextField search;
+
+    @Inject
+    private IndexService indexService;
+    @Inject
+    private int remoteDebounceTimeout = 200;
+
+    private final IntegerProperty pendingRequests = new SimpleIntegerProperty(0);
+
+    private boolean subscriptionRunning;
+
+    @Override
+    public void initialize(URL location, ResourceBundle resources) {
+        loader.managedProperty().bind(loader.visibleProperty());
+        loader.visibleProperty().bind(pendingRequests.greaterThan(0));
+
+        resultContainer.managedProperty().bind(resultContainer.visibleProperty());
+        resultContainer.visibleProperty().bind(loader.visibleProperty().not());
+
+        Observable.<String>create(subscriber -> {
+            search.textProperty().addListener((observable, oldValue, newValue) -> {
+                if (oldValue.equals(newValue)) {
+                    return;
+                }
+                try {
+                    String formattedNumber = formatPhoneNumber(newValue);
+                    if (!newValue.equals(formattedNumber)) {
+                        search.setText(formattedNumber);
+                        return;
+                    }
+                } catch (NumberParseException ignored) {}
+
+                subscriber.onNext(newValue);
+            });
+            subscriptionRunning = true;
+        })
+        .subscribeOn(Schedulers.computation())
+        .filter(this::isValidInput)
+        .debounce(remoteDebounceTimeout, TimeUnit.MILLISECONDS)
+        .observeOn(Schedulers.io())
+        .map(this::search)
+        .filter(list -> !list.isEmpty())
+        .observeOn(Schedulers.computation())
+        .subscribe(contacts -> {
+            Platform.runLater(resultContainer.getChildren()::clear);
+            contacts.forEach(this::showContact);
+        });
+    }
+
+    boolean isSubscriptionRunning() {
+        return subscriptionRunning;
+    }
+
+    private boolean isValidInput(String text) {
+        return !text.isEmpty() && (isValidPhoneNumber(text) || isValidEmail(text));
+    }
+
+    private boolean isValidEmail(String text) {
+        return text.length() > 3 && text.contains("@");
+    }
+
+    private void showContact(Contact contact) {
+        new IndexSearchResultItemView(contact).place(resultContainer);
+    }
+
+    private void incrementPendingRequests() {
+        synchronized (pendingRequests) {
+            pendingRequests.set(pendingRequests.get()+1);
+        }
+    }
+
+    private void decrementPendingRequests() {
+        synchronized (pendingRequests) {
+            pendingRequests.set(pendingRequests.get()-1);
+        }
+    }
+
+    private List<Contact> search(String searchText) {
+        incrementPendingRequests();
```

Comment:
Couldn't you just cancel other pending requests? Is there a need to keep "old" requests around?